lact-daemon/src/server/handler.rs

```diff
@@ -1055,72 +1058,81 @@ pub(crate) fn read_pci_db() -> Database {
     })
 }
 
-/// `sysfs_only` disables initialization of any external data sources, such as libdrm and nvml
-fn load_controllers(
-    base_path: &Path,
-    pci_db: &Database,
-) -> anyhow::Result<BTreeMap<String, DynGpuController>> {
-    let mut controllers = BTreeMap::new();
+#[cfg(any(test, not(feature = "nvidia")))]
+static NVML: LazyLock<Option<NvidiaLibs>> = LazyLock::new(|| None);
 
-    #[cfg(all(not(test), feature = "nvidia"))]
-    let nvml: LazyCell<Option<NvidiaLibs>> = LazyCell::new(|| match Nvml::init() {
-        Ok(nvml) => {
-            use crate::server::gpu_controller::NvApi;
+#[cfg(all(not(test), feature = "nvidia"))]
+// SAFETY: We use global LazyLock to make sure it's safe.
+#[allow(unused_unsafe)]
+static NVML: LazyLock<Option<NvidiaLibs>> = LazyLock::new(|| match unsafe { Nvml::init() } {
+    Ok(nvml) => {
+        use crate::server::gpu_controller::NvApi;
+
+        // The config has to be re-read here, because a LazyLock cannot capture external variables into the init closure
+        let disable_nvapi = Config::load()
+            .ok()
+            .flatten()
+            .and_then(|config| config.daemon.disable_nvapi);
 
-            // The config has to be re-read here, because a LazyCell cannot capture external variables into the init closure
-            let disable_nvapi = Config::load()
+        info!("Nvidia management library loaded");
+        let nvapi = if disable_nvapi == Some(true) {
+            info!("NvAPI support is disabled");
+            None
+        } else {
+            NvApi::new()
+                .inspect(|_| {
+                    info!("NvAPI library loaded");
+                })
+                .inspect_err(|err| {
+                    error!("could not load NvAPI library: {err:#}");
+                })
                 .ok()
-                .flatten()
-                .and_then(|config| config.daemon.disable_nvapi);
+        };
 
-            info!("Nvidia management library loaded");
-            let nvapi = if disable_nvapi == Some(true) {
-                info!("NvAPI support is disabled");
-                None
-            } else {
-                NvApi::new()
-                    .inspect(|_| {
-                        info!("NvAPI library loaded");
-                    })
-                    .inspect_err(|err| {
-                        error!("could not load NvAPI library: {err:#}");
-                    })
-                    .ok()
-            };
+        Some((Arc::new(nvml), Arc::new(nvapi)))
+    }
+    Err(err) => {
+        error!("could not load Nvidia management library: {err}");
+        None
+    }
+});
 
-            Some((Rc::new(nvml), Rc::new(nvapi)))
+static AMD_DRM: LazyLock<Option<LibDrmAmdgpu>> = LazyLock::new(|| {
+    // SAFETY: We use global LazyLock to make sure it's safe.
+    #[allow(unused_unsafe)]
+    // author of LibDrmAmdgpu refuses acknowledge it's unsafe: https://github.com/Umio-Yasuno/libdrm-amdgpu-sys-rs/issues/12
```

Comment:
Linking the libloading note would be useful: https://docs.rs/libloading/latest/libloading/struct.Library.html#thread-safety

And please avoid harassing the libdrm-amdgpu dev, their library has been extremely useful to LACT and they've contributed to the project in the past. It might be academically correct to mark the loader as unsafe, but it is indeed a situation where you are unlikely to get UB under real scenarios outside of parallel tests